File: src/flutter/dart/lib/ViewFactory_flutter.dart
/*
  This file is part of KDDockWidgets.

  SPDX-FileCopyrightText: 2019-2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
  Author: Sérgio Martins <sergio.martins@kdab.com>

  SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only

  Contact KDAB at <info@kdab.com> for commercial licensing options.
*/

import 'dart:ffi';

import 'package:KDDockWidgets/FloatingWindow_flutter.dart';
import 'package:KDDockWidgets/View_mixin.dart';
import 'package:KDDockWidgetsBindings/Bindings.dart' as KDDockWidgetBindings;
import 'DockWidget_flutter.dart';
import 'DropArea_flutter.dart';
import 'Group_flutter.dart';
import 'Separator_flutter.dart';
import 'Stack_flutter.dart';
import 'TitleBar_flutter.dart';
import 'TabBar_flutter.dart';

class ViewFactory_flutter extends KDDockWidgetBindings.ViewFactory_flutter {
  ViewFactory_flutter() {
    print("Dart: Created ViewFactory_flutter");
  }

  @override
  KDDockWidgetBindings.View createDropArea(
      KDDockWidgetBindings.DropArea? controller,
      KDDockWidgetBindings.View? parent) {
    print("ViewFactory_flutter: Creating DropArea");
    var da = DropArea_flutter(
        controller, KDDockWidgetBindings.KDDockWidgets_Type.DropArea, parent);
    print("ViewFactory_flutter: Created DropArea");
    return da;
  }

  @override
  KDDockWidgetBindings.ClassicIndicatorWindowViewInterface
      createClassicIndicatorWindow(
          KDDockWidgetBindings.ClassicIndicators? indicators) {
    print("ViewFactory_flutter: createClassicIndicatorWindow");
    return KDDockWidgetBindings.ClassicIndicatorWindowViewInterface();
  }

  @override
  KDDockWidgetBindings.View createGroup(KDDockWidgetBindings.Group? group,
      {required KDDockWidgetBindings.View? parent}) {
    print("ViewFactory_flutter: createGroup");
    return Group_flutter(group, parent);
  }

  @override
  KDDockWidgetBindings.View createTabBar(KDDockWidgetBindings.TabBar? tabBar,
      {required KDDockWidgetBindings.View? parent}) {
    print("ViewFactory_flutter: createTabBar");
    return TabBar_flutter(tabBar, parent);
  }

  @override
  KDDockWidgetBindings.View createTitleBar(
      KDDockWidgetBindings.TitleBar? controller,
      KDDockWidgetBindings.View? parent) {
    print("ViewFactory_flutter: createTitleBar");
    return TitleBar_flutter(controller, parent);
  }

  @override
  KDDockWidgetBindings.View createStack(
      KDDockWidgetBindings.Stack? stack, KDDockWidgetBindings.View? parent) {
    print("ViewFactory_flutter: createStack");
    return Stack_flutter(stack, parent);
  }

  @override
  KDDockWidgetBindings.View createSeparator(
      KDDockWidgetBindings.Separator? separator,
      {required KDDockWidgetBindings.View? parent}) {
    return Separator_flutter(separator, parent: parent);
  }

  @override
  KDDockWidgetBindings.View createDockWidget(String? uniqueName,
      {int options = 0, int layoutSaverOptions = 0, int windowFlags = 0}) {
    return DockWidget_flutter(uniqueName,
        options: options, layoutSaverOptions: layoutSaverOptions);
  }

  @override
  KDDockWidgetBindings.View createFloatingWindow(
      KDDockWidgetBindings.FloatingWindow? controller,
      {required KDDockWidgetBindings.MainWindow? parent,
      int windowFlags = 0}) {
    // TODOm4: What to do with main window pointer
    return FloatingWindow_flutter(controller, null);
  }
}

File: src/flutter/ViewFactory_flutter.cpp
/*
  This file is part of KDDockWidgets.

  SPDX-FileCopyrightText: 2019-2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
  Author: Sérgio Martins <sergio.martins@kdab.com>

  SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only

  Contact KDAB at <info@kdab.com> for commercial licensing options.
*/

#include "ViewFactory_flutter.h"
#include "Config.h"

#include "core/Utils_p.h"

#include "kddockwidgets/core/TabBar.h"
#include "kddockwidgets/core/Stack.h"
#include "kddockwidgets/core/FloatingWindow.h"
#include "kddockwidgets/core/indicators/ClassicIndicators.h"
#include "kddockwidgets/core/indicators/NullIndicators.h"
#include "kddockwidgets/core/indicators/SegmentedIndicators.h"
#include "kddockwidgets/core/MainWindow.h"

// #include "qtwidgets/views/ClassicIndicatorsWindow.h"
// #include "qtwidgets/views/SegmentedIndicatorsOverlay.h"
// #include "qtwidgets/views/FloatingWindow.h"
// #include "qtwidgets/views/DockWidget.h"
// #include "qtwidgets/views/DropArea.h"
// #include "qtwidgets/views/Group.h"
// #include "qtwidgets/views/View.h"
// #include "qtwidgets/views/Separator.h"
// #include "qtwidgets/views/TitleBar.h"
// #include "qtwidgets/views/TabBar.h"
// #include "qtwidgets/views/SideBar.h"
// #include "qtwidgets/views/Stack.h"
// #include "qtwidgets/views/MainWindow.h"
// #include "qtwidgets/views/MDILayout.h"
// #include "qtwidgets/views/RubberBand.h"

#include "views/ClassicIndicatorWindowViewInterface.h"


// clazy:excludeall=ctor-missing-parent-argument

using namespace KDDockWidgets;
using namespace KDDockWidgets::flutter;

ViewFactory_flutter::~ViewFactory_flutter()
{
}

Core::View *ViewFactory_flutter::createDockWidget(const QString &, DockWidgetOptions, LayoutSaverOptions,
                                                  Qt::WindowFlags) const
{
    return {};
}


Core::View *ViewFactory_flutter::createGroup(Core::Group *, Core::View *) const
{
    Q_ASSERT(false);
    return {};
}

Core::View *ViewFactory_flutter::createTitleBar(Core::TitleBar *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory_flutter::createTabBar(Core::TabBar *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory_flutter::createStack(Core::Stack *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory_flutter::createSeparator(Core::Separator *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory_flutter::createFloatingWindow(Core::FloatingWindow *,
                                                      Core::MainWindow *, Qt::WindowFlags) const
{
    return {};
}

Core::View *ViewFactory_flutter::createRubberBand(Core::View *) const
{
    return nullptr;
}

Core::View *ViewFactory_flutter::createSideBar(Core::SideBar *, Core::View *) const
{
    return {};
}

QAbstractButton *ViewFactory_flutter::createTitleBarButton(QWidget *, TitleBarButtonType) const
{
    return nullptr;
}

// iconForButtonType impl is the same for QtQuick and QtWidgets
Icon ViewFactory_flutter::iconForButtonType(TitleBarButtonType, qreal) const
{
    return {};
}

Core::View *ViewFactory_flutter::createDropArea(Core::DropArea *, Core::View *) const
{
    return {};
}

Core::View *ViewFactory_flutter::createMDILayout(Core::MDILayout *, Core::View *) const
{
    return {};
}

Core::View *
ViewFactory_flutter::createSegmentedDropIndicatorOverlayView(Core::SegmentedIndicators *,
                                                             Core::View *) const
{
    return {};
}

Core::ClassicIndicatorWindowViewInterface *
ViewFactory_flutter::createClassicIndicatorWindow(Core::ClassicIndicators *) const
{
    return {};
}
